File: plugins/codemodder-plugin-maven/src/main/java/io/codemodder/plugins/maven/operator/AbstractCommand.java
package io.codemodder.plugins.maven.operator;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import javax.xml.stream.XMLStreamException;
import org.dom4j.Element;
import org.dom4j.Node;

/** Base implementation of Command - used by SimpleDependency and SimpleInsert */
abstract class AbstractCommand implements Command {

  /**
   * Given a POM, locate its coordinates for a given dependency based on lookupExpression and
   * determines whether an upgrade is necessary.
   *
   * @param pm The ProjectModel containing the POM and other project information.
   * @param lookupExpression An XPath expression to locate the desired dependency node.
   * @return true if an upgrade was performed, false otherwise.
   */
  protected boolean handleDependency(ProjectModel pm, String lookupExpression) {
    List<Node> dependencyNodes =
        Util.selectXPathNodes(pm.getPomFile().getResultPom(), lookupExpression);

    if (1 == dependencyNodes.size()) {
      List<Node> versionNodes = Util.selectXPathNodes(dependencyNodes.get(0), "./m:version");

      if (1 == versionNodes.size()) {
        Element versionNode = (Element) versionNodes.get(0);

        boolean mustUpgrade = true;

        if (pm.isSkipIfNewer()) {
          mustUpgrade = Util.findOutIfUpgradeIsNeeded(pm, versionNode);
        }

        if (mustUpgrade) {
          Util.upgradeVersionNode(pm, versionNode, pm.getPomFile());
        }

        return true;
      }
    }

    return false;
  }

  @Override
  public boolean execute(ProjectModel pm)
      throws URISyntaxException, IOException, XMLStreamException {
    return false;
  }

  @Override
  public boolean postProcess(ProjectModel c) throws XMLStreamException {
    return false;
  }

  /**
   * Retrieves the local repository path from various sources or default values.
   *
   * @param pm The ProjectModel containing information about the project.
   * @return A File object representing the local repository path.
   */
  protected Path getLocalRepositoryPath(ProjectModel pm) {
    Path localRepositoryPath = null;

    if (pm.getRepositoryPath() != null) {
      localRepositoryPath = pm.getRepositoryPath();
    } else if (System.getenv("M2_REPO") != null) {
      localRepositoryPath = Paths.get(System.getenv("M2_REPO"));
    } else if (System.getProperty("maven.repo.local") != null) {
      localRepositoryPath = Paths.get(System.getProperty("maven.repo.local"));
    } else {
      localRepositoryPath = Paths.get(System.getProperty("user.home"), ".m2/repository");
    }

    return localRepositoryPath;
  }
}

File: pom-operator/src/main/java/io/github/pixee/maven/operator/AbstractCommand.kt
package io.github.pixee.maven.operator

import io.github.pixee.maven.operator.Util.findOutIfUpgradeIsNeeded
import io.github.pixee.maven.operator.Util.selectXPathNodes
import io.github.pixee.maven.operator.Util.upgradeVersionNode
import org.dom4j.Element
import java.io.File

/**
 * Base implementation of Command - used by SimpleDependency and SimpleInsert
 */
abstract class AbstractCommand : Command {
    /**
     * Given a POM, locate its coordinates for a given dependency based on lookupExpression and figures out the upgrade
     *
     * TODO review this
     */
    protected fun handleDependency(pm: ProjectModel, lookupExpression: String): Boolean {
        val dependencyNodes = pm.pomFile.resultPom.selectXPathNodes(lookupExpression)

        if (1 == dependencyNodes.size) {
            val versionNodes = dependencyNodes[0].selectXPathNodes("./m:version")

            if (1 == versionNodes.size) {
                val versionNode = versionNodes[0] as Element

                var mustUpgrade = true

                if (pm.skipIfNewer) {
                    mustUpgrade = findOutIfUpgradeIsNeeded(pm, versionNode)
                }

                if (mustUpgrade) {
                    upgradeVersionNode(pm, versionNode, pm.pomFile)
                }

                return true
            }
        }

        return false
    }

    override fun execute(pm: ProjectModel): Boolean = false

    override fun postProcess(c: ProjectModel): Boolean = false

    protected fun getLocalRepositoryPath(pm: ProjectModel): File {
        val localRepositoryPath: File = when {
            pm.repositoryPath != null -> pm.repositoryPath
            System.getenv("M2_REPO") != null -> File(System.getenv("M2_REPO"))
            System.getProperty("maven.repo.local") != null -> File(System.getProperty("maven.repo.local"))
            else -> File(
                System.getProperty("user.home"),
                ".m2/repository"
            )
        }

        return localRepositoryPath
    }
}
